sabnzbd/downloader.py

```diff
@@ -771,49 +747,44 @@ def run(self):
                 continue
 
             if self.recv_threads > 1:
-                for nw, bytes_received, done in self.recv_pool.map(self.__recv, read):
-                    if bytes_received > last_max_chunk_size:
-                        last_max_chunk_size = bytes_received
-                    self.__handle_recv_result(nw, bytes_received, done)
-                if self.bandwidth_limit:
-                    self.__check_speed()
+                # Submit a process_nw task to the pool for every NewWrapper which is readable
+                for selected in read:
+                    if nw := self.read_fds.get(selected):
+                        self.process_tasks.add(self.recv_pool.submit(self.process_nw, nw))
+
+                # Wait for all the tasks to complete
+                concurrent.futures.wait(self.process_tasks)
+
+                # Clear task list
+                self.process_tasks.clear()
             else:
                 for selected in read:
-                    nw, bytes_received, done = self.__recv(selected)
-                    if bytes_received > last_max_chunk_size:
-                        last_max_chunk_size = bytes_received
-                    self.__handle_recv_result(nw, bytes_received, done)
-                    if self.bandwidth_limit and bytes_received:
-                        self.__check_speed()
-
-    def __recv(self, selected):
-        nw = None
+                    if nw := self.read_fds.get(selected):
+                        self.process_nw(nw)
+
+            self.__check_speed()
+            self.__check_assembler()
+
+    def process_nw(self, nw: NewsWrapper) -> None:
+        """Receive data from NewsWrapper and handle response"""
         try:
-            nw = self.read_fds[selected]
             bytes_received, done = nw.recv_chunk()
-            return nw, bytes_received, done
         except ssl.SSLWantReadError:
-            return nw, 0, False
+            return
         except:
-            return nw, 0, True
-
-    def __check_speed(self):
-        BPSMeter = sabnzbd.BPSMeter
-        if BPSMeter.bps + BPSMeter.sum_cached_amount > self.bandwidth_limit:
-            BPSMeter.update()
-            while BPSMeter.bps > self.bandwidth_limit:
-                time.sleep(0.01)
-                BPSMeter.update()
-
-    def __handle_recv_result(self, nw: NewsWrapper, bytes_received: int = 0, done: bool = False):
-        if not bytes_received:
-            if done:
-                self.__reset_nw(nw, "server closed connection", wait=False)
+            self.__reset_nw(nw, "server closed connection", wait=False)
             return
 
         article = nw.article
         server = nw.server
-        sabnzbd.BPSMeter.update(server.id, bytes_received)
+
+        with sabnzbd.decorators.NZBQUEUE_LOCK:
```

Comment:
I think we need a different lock for this.
The NZBQUEUE_LOCK is meant for things that interact with the queue, like getting articles or removing NZO's. 
Seems we just need a lock "local" to the Downloader for these things?